xarray/core/missing.py

```diff
@@ -365,26 +515,112 @@ def interp_na(
         arr = apply_ufunc(
             interpolator,
             self,
-            index,
+            index.values,
             input_core_dims=[[dim], [dim]],
             output_core_dims=[[dim]],
             output_dtypes=[self.dtype],
             dask="parallelized",
             vectorize=True,
             keep_attrs=keep_attrs,
         ).transpose(*self.dims)
+    return arr
 
-    if limit is not None:
-        arr = arr.where(valids)
 
-    if max_gap is not None:
-        if dim not in self.coords:
-            raise NotImplementedError(
-                "max_gap not implemented for unlabeled coordinates yet."
-            )
-        nan_block_lengths = _get_nan_block_lengths(self, dim, index)
-        arr = arr.where(nan_block_lengths <= max_gap)
+class GapMask:
+    """An object that allows for flexible masking of gaps."""
+
+    def __init__(self, content: DataArray | Dataset, mask: np.ndarray):
+        self.content = content
+        self.mask = mask
 
+    def _fill_method(name: str, _fill_function: Callable | None = None):
+        def method(self, *args, _fill_function=_fill_function, **kwargs):
+            if _fill_function is None:
+                _fill_function = getattr(self.content, name)
+                filled = _fill_function(*args, **kwargs)
+            else:
+                filled = _fill_function(self.content, *args, **kwargs)
+
+            if self.mask is not None:
+                filled = filled.where(~self.mask, other=self.content)
+            return filled
+
+        method.__name__ = name
+        method.__doc__ = _FILL_MISSING_DOCSTRING_TEMPLATE.format(name=name)
+        return method
+
+    ffill = _fill_method("ffill")
+    bfill = _fill_method("bfill")
+    fillna = _fill_method("fillna")
+    interpolate_na = _fill_method("interpolate_na")
+
+
+def mask_gaps(
+    self,
```

Comment:
```suggestion
    self: T_Xarray,
```
Since this is not a member method of any class, consider naming this `obj` instead.

Reply:
True, I renamed all self outside classes!